File: android/app/src/test/java/miko/todayinspacehistory/support/MockNetworkClient.kt
package miko.todayinspacehistory.support

import miko.todayinspacehistory.data.Errors
import miko.todayinspacehistory.data.model.ApiResponse
import miko.todayinspacehistory.data.network.NetworkClient

class MockNetworkClient : NetworkClient {
    var searchResult: Result<ApiResponse> = Result.failure(Errors.ImageProvider.NoItems)
    var fetchImagesResult: Result<List<String>> = Result.success(emptyList())
    private val _fetchImagesUrls = mutableListOf<String>()
    val fetchImagesUrls: List<String>
        get() = _fetchImagesUrls.toList()

    override suspend fun search(): ApiResponse {
        return searchResult.getOrThrow()
    }

    override suspend fun fetchImageUrls(url: String): List<String> {
        _fetchImagesUrls += url
        return fetchImagesResult.getOrThrow()
    }
}

File: ios/TodayInSpaceHistoryTests/Support/MockNetworkClient.swift
//
//  MockNetworkClient.swift
//  TodayInSpaceHistoryTests
//

import Foundation
@testable import TodayInSpaceHistory

final class MockNetworkClient: NetworkClientProtocol {
    var searchResult: Result<APIResponse, Error> = .failure(Errors.ImageProvider.noItems)
    var fetchImagesResult: Result<[String], Error> = .success([])
    private(set) var fetchImagesURLs: [String] = []

    func makeRequest<T: Decodable>(endpoint: Endpoints) async throws -> T {
        let response = try searchResult.get()
        guard let typed = response as? T else {
            throw Errors.NetworkClient.decodingFailed(underlying: NSError(domain: "MockNetworkClient", code: 1))
        }
        return typed
    }

    func fetchImages<T: Decodable>(url: String) async throws -> T {
        fetchImagesURLs.append(url)
        let urls = try fetchImagesResult.get()
        guard let typed = urls as? T else {
            throw Errors.NetworkClient.decodingFailed(underlying: NSError(domain: "MockNetworkClient", code: 2))
        }
        return typed
    }
}
